File: RetroDRYClient/unittest/datonKeyTests.ts
import { equal } from 'assert';
import 'mocha';
import DatonKey, {parseDatonKey} from '../src/datonKey';

describe('daton key', () => {

    it('should parse', () => {
        //new persiston
        let parsedKey = parseDatonKey('Customer|=-1');
        equal(parsedKey.isNew(), true);
        equal(parsedKey.isPersiston(), true);
        equal(parsedKey.typeName, 'Customer');
        equal(parsedKey.otherSegments.length, 1);
        equal(parsedKey.otherSegments[0], '=-1');

        //existing persiston
        parsedKey = parseDatonKey('Customer|=67');
        equal(parsedKey.isNew(), false);
        equal(parsedKey.isPersiston(), true);
        equal(parsedKey.typeName, 'Customer');
        equal(parsedKey.otherSegments.length, 1);
        equal(parsedKey.otherSegments[0], '=67');
        equal(parsedKey.persistonKeyAsInt(), 67);
        equal(parsedKey.persistonKeyAsString(), '67');

        //whole table persiston
        parsedKey = parseDatonKey('CustomerType|+');
        equal(parsedKey.isNew(), false);
        equal(parsedKey.isPersiston(), true);
        equal(parsedKey.typeName, 'CustomerType');
        equal(parsedKey.otherSegments.length, 1);
        equal(parsedKey.otherSegments[0], '+');

        //viewon without criteria
        parsedKey = parseDatonKey('CustomerList');
        equal(parsedKey.isNew(), false);
        equal(parsedKey.isPersiston(), false);
        equal(parsedKey.typeName, 'CustomerList');
        equal(parsedKey.otherSegments.length, 0);

        //viewon with criteria
        parsedKey = parseDatonKey('CustomerList|code1=\\\\/|code2=\\||code3=a=\\|b\\|');
        equal(parsedKey.isNew(), false);
        equal(parsedKey.isPersiston(), false);
        equal(parsedKey.typeName, 'CustomerList');
        equal(parsedKey.otherSegments.length, 3);
        equal(parsedKey.otherSegments[0], 'code1=\\/');
        equal(parsedKey.otherSegments[1], 'code2=|');
        equal(parsedKey.otherSegments[2], 'code3=a=|b|');
    });

    it('should escape segment in isolation', () => {
        let k = new DatonKey('ItemList', []);
        equal(k.escapeSegment('x'), 'x');
    });

    it('should escape segments', () => {
        let k = new DatonKey('CustomerList', ['code1=\\/', 'code3=a=|b|', 'code2=|', '_sort=firstName', '_page=3']);
        equal(k.toKeyString(), 'CustomerList|_page=3|_sort=firstName|code1=\\\\/|code2=\\||code3=a=\\|b\\|');
    });
});

File: ServerUnitTest/DatonKeyTests.cs
using System;
using System.Linq;
using Microsoft.VisualStudio.TestTools.UnitTesting;
using RetroDRY;

namespace UnitTest
{
    [TestClass]
    public class DatonKeyTests
    {
        [TestMethod]
        public void FormatAsString()
        {
            DatonKey k = new PersistonKey("Emp", "3", false);
            Assert.AreEqual("Emp|=3", k.ToString());

            k = new PersistonKey("Emp", "3", true);
            Assert.AreEqual("Emp|+", k.ToString());

            var criteria = new[]
            {
                new ViewonKey.Criterion("Region", "NE"),
                new ViewonKey.Criterion("Funny", "1")
            };
            k = new ViewonKey("EmpList", criteria, sortColumnName: "LastName", pageNo: 2);
            Assert.AreEqual("EmpList|_page=2|_sort=LastName|Funny=1|Region=NE", k.ToString());

            //now with crazy characters 
            criteria = new[]
            {
                new ViewonKey.Criterion("Crazy", "Angry \"O'Leary\" :| :\\"), //actual: Angry "O'Leary" :| :\
                new ViewonKey.Criterion("Zany", "a=|b|"),
                new ViewonKey.Criterion("Loopy", "\\/") //actual: \/
            };
            k = new ViewonKey("EmpList", criteria);
            Assert.AreEqual(@"EmpList|Crazy=Angry ""O'Leary"" :\| :\\|Loopy=\\/|Zany=a=\|b\|", k.ToString());
        }

        [TestMethod]
        public void Parse()
        { 
            //existing persiston
            var pk = DatonKey.Parse("Emp|=3") as PersistonKey;
            Assert.IsNotNull(pk);
            Assert.AreEqual("Emp", pk.Name);
            Assert.AreEqual("3", pk.PrimaryKey);
            Assert.IsFalse(pk.WholeTable);
            Assert.IsFalse(pk.IsNew);

            //new persiston
            pk = DatonKey.Parse("Emp|=-1") as PersistonKey;
            Assert.IsNotNull(pk);
            Assert.AreEqual("Emp", pk.Name);
            Assert.AreEqual("-1", pk.PrimaryKey);
            Assert.IsFalse(pk.WholeTable);
            Assert.IsTrue(pk.IsNew);

            //whole table persiston
            pk = DatonKey.Parse("EmpType|+") as PersistonKey;
            Assert.IsNotNull(pk);
            Assert.AreEqual("EmpType", pk.Name);
            Assert.IsTrue(pk.WholeTable);
            Assert.IsFalse(pk.IsNew);

            //viewon without criteria
            var vk = DatonKey.Parse("EmpList") as ViewonKey;
            Assert.IsNotNull(vk?.Criteria);
            Assert.AreEqual("EmpList", vk.Name);
            Assert.AreEqual(0, vk.PageNumber);
            Assert.IsNull(vk.SortColumnName);
            Assert.AreEqual(0, vk.Criteria.Count());

            //viewon with escaped criteria
            vk = DatonKey.Parse(@"CustomerList|code1=\\/|code2=\||code3=a=\|b\|") as ViewonKey;
            Assert.IsNotNull(vk?.Criteria);
            Assert.AreEqual("CustomerList", vk.Name);
            Assert.AreEqual(3, vk.Criteria.Count());
            var criArray = vk.Criteria.ToArray();
            Assert.AreEqual("code1", criArray[0].Name);
            Assert.AreEqual(@"\/", criArray[0].PackedValue);
            Assert.AreEqual("code2", criArray[1].Name);
            Assert.AreEqual("|", criArray[1].PackedValue);
            Assert.AreEqual("code3", criArray[2].Name);
            Assert.AreEqual("a=|b|", criArray[2].PackedValue);

            //with sort and page
            vk = DatonKey.Parse("EmpList|_page=2|_sort=firstName") as ViewonKey;
            Assert.IsNotNull(vk?.Criteria);
            Assert.AreEqual("EmpList", vk.Name);
            Assert.IsNotNull(vk);
            Assert.AreEqual(2, vk.PageNumber);
            Assert.AreEqual("firstName", vk.SortColumnName);
            Assert.AreEqual(0, vk.Criteria.Count());
        }
    }
}
